Yank/multistate/multistateanalyzer.py

```diff
@@ -612,18 +612,18 @@ def n_states(self):
         return self._reporter.n_states
 
     def _get_end_thermodynamic_states(self):
-        """Read thermodynamic states at the ends of the protocol."""
-        # TODO: what if reference_states changes? This should become a CachedProperty.
-        # Check cached values.
-        if self._end_thermodynamic_states is None:
-            sampled_states, unsampled_states = self._reporter.read_thermodynamic_states()
-            if len(unsampled_states) == 0:
-                self._end_thermodynamic_states = [sampled_states[0], sampled_states[-1]]
-            else:
-                self._end_thermodynamic_states = unsampled_states
-            # Cache other useful informations since we have already read this.
-            # TODO should we read temperatures of all the states and let kT property depend on reference_states?
-            self._kT = self._end_thermodynamic_states[0].kT
+        """Read thermodynamic states at the ends of the protocol.
+
+        Returns
+        -------
+        end_thermodynamic_states : list of ThermodynamicState
+            The end thermodynamic states
+
+        """
+        self._end_thermodynamic_states = self._reporter.read_end_thermodynamic_states()
```

Comment:
Hmm. Why removing the `if self._end_thermodynamic_states is None` here? This will read twice the states during a normal analysis with restraint unbiasing.

Reply:
Good catch! That was unintentional. I'll fix in the next PR.